styx-scheduler-service/src/main/java/com/spotify/styx/Scheduler.java

```diff
@@ -159,17 +173,44 @@ void tick() {
 
     timedOutInstances.forEach(this::sendTimeout);
 
-    for (InstanceState eligibleInstance : eligibleInstances) {
-      final boolean proceed = limitAndDequeue(
-          resources, workflowResourceReferences, currentResourceUsage, eligibleInstance);
-      if (!proceed) {
-        break;
-      }
-    }
+    limitAndDequeue(config, resources, workflowResourceReferences, currentResourceUsage,
+        eligibleInstances);
 
     updateStats(resources, currentResourceUsage);
   }
 
+  private void limitAndDequeue(
+      StyxConfig config,
+      Map<String, Resource> resources,
+      Map<WorkflowId, Set<String>> workflowResourceReferences,
+      Map<String, Long> currentResourceUsage, List<InstanceState> eligibleInstances) {
+
+    // Enable gating unless disabled in runtime config
+    final WorkflowExecutionGate gate = config.executionGatingEnabled() ? this.gate : NOOP;
+
+    // Process the eligible instances in batches in order to parallelize execution blocker lookup
+    for (List<InstanceState> batch : Lists.partition(eligibleInstances, SCHEDULING_BATCH_SIZE)) {
+
+      // Asynchronously look up execution blockers for a batch of instances
+      final Map<WorkflowInstance, CompletionStage<Optional<ExecutionBlocker>>> executionBlockers =
+          batch.stream()
+              .map(InstanceState::workflowInstance)
+              .collect(toMap(identity(), gate::executionBlocker));
+
+      // Evaluate each instance in the batch for dequeuing
+      for (InstanceState instance : batch) {
+        final boolean proceed = limitAndDequeue(
+            resources, workflowResourceReferences, currentResourceUsage, instance,
+            executionBlockers.getOrDefault(instance.workflowInstance(), NO_BLOCKER));
+
+        // Stop processing if rate limit was hit
```

Comment:
First of all this comment is now lying. Second, if one blocker execution is interrupted, are we sure we want to stop processing everything left? I guess in many cases interruption would mean the executor service is shutting down.

Reply:
Yeah I think we want to stop processing if the thread was interrupted.

Reply:
+1